### DISCUSSION TITLE
Successfully deploying to Preview, failing to deploy to Production, getting HTML instead of JSON on get request

### DISCUSSION THREAD

**Discussion by TheoKondak** [vercel/community#1975] [Help] [answered]
## TLDR

When I deploy my NextJs app to Preview, all `GET` requests to my local endpoints work, when I deploy to Production, an api route `api/getSettings` returns 404.

So:
`theokondak.com/api/getSettings` Does not work
`dev.theokondak.com/api/getSettings` Works

### Summary

I have a personal project in (NextJs, Sanity), that I am working on, and I am facing some issues when trying to deploy it to Production.

The deployment was working fine,  but recently I have added some new features and wanted to publish them to Production.
Essentially the new features include a new request to Sanity, to fetch some data. The data fetch works:
- Sanity Studio Vision
- localhost
- Preview deployment environment

But as already mentioned **it doesn't work on the Production**.

I am deploying the project using the Github integration of Vercel. The project has two main branches:
- main: Production (doesn't deploy)
- staging: Preview (deploys successfully)

I have checked the environment variables on Vercel and they seem OK (besides there are several other requests that are being fulfilled successfully already).

## This is the error that I am getting:

```
[12:09:07.888] Running build in Cleveland, USA (East) – cle1
[12:09:07.934] Cloning github.com/TheoKondak/portfolio (Branch: main, Commit: a05c13d)
[12:09:08.774] Cloning completed: 839.592ms
[12:09:11.783] Restored build cache
[12:09:11.814] Running "vercel build"
[12:09:12.381] Vercel CLI 28.18.3
[12:09:12.769] Installing dependencies...
[12:09:13.135] yarn install v1.22.17
[12:09:13.170] info No lockfile found.
[12:09:13.180] [1/4] Resolving packages...
[12:09:18.194] warning icons@1.0.0: Package no longer supported. Contact support@npmjs.com for more info.
[12:09:23.403] [2/4] Fetching packages...
[12:09:49.855] [3/4] Linking dependencies...
[12:09:49.859] warning " > next-sanity@4.1.6" has unmet peer dependency "@sanity/icons@2".
[12:09:49.859] warning " > next-sanity@4.1.6" has unmet peer dependency "@sanity/types@3".
[12:09:49.859] warning " > next-sanity@4.1.6" has unmet peer dependency "@sanity/ui@1".
[12:09:49.860] warning " > next-sanity@4.1.6" has unmet peer dependency "sanity@3".
[12:09:49.860] warning " > next-sitemap@4.0.6" has unmet peer dependency "@next/env@*".
[12:09:49.861] warning " > styled-components@5.3.9" has unmet peer dependency "react-is@>= 16.8.0".
[12:10:05.190] [4/4] Building fresh packages...
[12:10:05.247] success Saved lockfile.
[12:10:05.255] Done in 52.13s.
[12:10:05.292] Detected Next.js version: 13.1.6
[12:10:05.294] Running "yarn run build"
[12:10:05.512] yarn run v1.22.17
[12:10:05.551] $ next build
[12:10:06.257] info  - Linting and checking validity of types...
[12:10:12.608] info  - Creating an optimized production build...
[12:10:27.493] info  - Compiled successfully
[12:10:27.493] info  - Collecting page data...
[12:10:35.196] info  - Generating static pages (0/3)
[12:10:38.541] Requesting siteSettings using Axios
[12:10:38.541] Fetching: https://theokondak.com/api/getSettings
[12:10:38.541] ---
[12:10:38.558] ERROR OCCURED
[12:10:38.559] Error Response Data
[12:10:38.559] <!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>404: This page could not be found</title><meta name="next-head-count" content="3"/><meta name="description" content="Hello, my name is Theo, and I&#x27;m an aspiring developer who&#x27;s passionate about exploring the latest developments in AI, cloud computing, React, and more. Come and visit my portfolio to check out some of my projects and discover what else I&#x27;ve been working on!"/><meta property="og:title" content="MYNAME | Portfolio"/><meta property="og:description" content="Hello, my name is Theo, and I&#x27;m an aspiring developer who&#x27;s passionate about exploring the latest developments in AI, cloud computing, React, and more. Come and visit my portfolio to check out some of my projects and discover what else I&#x27;ve been working on!"/><meta property="og:image" content="https://theokondak.com/_next/image?url=https%3A%2F%2Fcdn.sanity.io%2Fimages%2Fchvg577r%2Fproduction%2Fb5275e2e2b3dd215b4d81e21d4efcb8fc0a0ba18-1024x1024.png&amp;w=640&amp;q=75"/><link rel="preload" href="/_next/static/css/3f963a6b249549cf.css" as="style"/><link rel="stylesheet" href="/_next/static/css/3f963a6b249549cf.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js"></script><script src="/_next/static/chunks/webpack-fa78fa874df28d6e.js" defer=""></script><script src="/_next/static/chunks/framework-2c79e2a64abdb08b.js" defer=""></script><script src="/_next/static/chunks/main-5861cb26bf5df503.js" defer=""></script><script src="/_next/static/chunks/pages/_app-891652dd44e1e4e1.js" defer=""></script><script src="/_next/static/chunks/pages/_error-8353112a01355ec2.js" defer=""></script><script src="/_next/static/f7ZeYtPMPIBuwsemTYQNI/_buildManifest.js" defer=""></script><script src="/_next/static/f7ZeYtPMPIBuwsemTYQNI/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div style="font-family:-apple-system, BlinkMacSystemFont, Roboto, &quot;Segoe UI&quot;, &quot;Fira Sans&quot;, Avenir, &quot;Helvetica Neue&quot;, &quot;Lucida Grande&quot;, sans-serif;height:100vh;text-align:center;display:flex;flex-direction:column;align-items:center;justify-content:center"><div><style>
[12:10:38.560]                 body { margin: 0; color: #000; background: #fff; }
[12:10:38.561]                 .next-error-h1 {
[12:10:38.561]                   border-right: 1px solid rgba(0, 0, 0, .3);
[12:10:38.561]                 }
[12:10:38.561] 
[12:10:38.561]                 @media (prefers-color-scheme: dark) {
[12:10:38.561]                   body { color: #fff; background: #000; }
[12:10:38.561]                   .next-error-h1 {
[12:10:38.561]                     border-right: 1px solid rgba(255, 255, 255, .3);
[12:10:38.562]                   }
[12:10:38.562]                 }</style><h1 class="next-error-h1" style="display:inline-block;margin:0;margin-right:20px;padding:0 23px 0 0;font-size:24px;font-weight:500;vertical-align:top;line-height:49px">404</h1><div style="display:inline-block;text-align:left;line-height:49px;height:49px;vertical-align:middle"><h2 style="font-size:14px;font-weight:normal;line-height:49px;margin:0;padding:0">This page could not be found<!-- -->.</h2></div></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"statusCode":404}},"page":"/_error","query":{},"buildId":"f7ZeYtPMPIBuwsemTYQNI","nextExport":true,"isFallback":false,"gip":true,"scriptLoader":[]}</script></body></html>
[12:10:38.563] ---
[12:10:38.563] Error Response Status
[12:10:38.563] 404
[12:10:38.563] ---
[12:10:38.563] Error Response Headers
[12:10:38.563] AxiosHeaders {
[12:10:38.563]   'access-control-allow-origin': '*',
[12:10:38.564]   age: '164250',
[12:10:38.564]   'cache-control': 'public, max-age=0, must-revalidate',
[12:10:38.564]   'content-disposition': 'inline; filename="404"',
[12:10:38.564]   'content-type': 'text/html; charset=utf-8',
[12:10:38.564]   date: 'Wed, 05 Apr 2023 10:10:38 GMT',
[12:10:38.564]   etag: 'W/"fe2e7334be1b992881807014f04124d4"',
[12:10:38.584]   server: 'Vercel',
[12:10:38.584]   'strict-transport-security': 'max-age=63072000',
[12:10:38.584]   'x-matched-path': '/404',
[12:10:38.585]   'x-vercel-cache': 'HIT',
[12:10:38.585]   'x-vercel-id': 'cle1:cle1::lcb4w-1680689438551-9fce2fff0760',
[12:10:38.585]   connection: 'close',
[12:10:38.585]   'transfer-encoding': 'chunked'
[12:10:38.585] }
[12:10:38.585] ---
[12:10:38.586] {
[12:10:38.586]   transitional: {
[12:10:38.586]     silentJSONParsing: true,
[12:10:38.587]     forcedJSONParsing: true,
[12:10:38.587]     clarifyTimeoutError: false
[12:10:38.587]   },
[12:10:38.587]   adapter: [ 'xhr', 'http' ],
[12:10:38.587]   transformRequest: [ [Function: transformRequest] ],
[12:10:38.587]   transformResponse: [ [Function: transformResponse] ],
[12:10:38.587]   timeout: 0,
[12:10:38.587]   xsrfCookieName: 'XSRF-TOKEN',
[12:10:38.588]   xsrfHeaderName: 'X-XSRF-TOKEN',
[12:10:38.588]   maxContentLength: -1,
[12:10:38.590]   maxBodyLength: -1,
[12:10:38.590]   env: {
[12:10:38.590]     FormData: [Function: FormData] {
[12:10:38.590]       LINE_BREAK: '\r\n',
[12:10:38.590]       DEFAULT_CONTENT_TYPE: 'application/octet-stream'
[12:10:38.590]     },
[12:10:38.590]     Blob: [class Blob]
[12:10:38.590]   },
[12:10:38.591]   validateStatus: [Function: validateStatus],
[12:10:38.591]   headers: AxiosHeaders {
[12:10:38.591]     Accept: 'application/json, text/plain, */*',
[12:10:38.591]     'User-Agent': 'axios/1.3.4',
[12:10:38.591]     'Accept-Encoding': 'gzip, compress, deflate, br'
[12:10:38.591]   },
[12:10:38.591]   method: 'get',
[12:10:38.591]   url: 'https://theokondak.com/api/getSettings',
[12:10:38.591]   data: undefined
[12:10:38.591] }
[12:10:38.591] info  - Generating static pages (3/3)
[12:10:38.591] 
[12:10:38.592] Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
[12:10:38.592] TypeError: Cannot read properties of undefined (reading 'settings')
[12:10:38.592]     at fetchSettings (/vercel/path0/.next/server/pages/index.js:1562:31)
[12:10:38.592]     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[12:10:38.592]     at async getStaticProps (/vercel/path0/.next/server/pages/index.js:1275:26)
[12:10:38.592]     at async renderToHTML (/vercel/path0/node_modules/next/dist/server/render.js:385:20)
[12:10:38.592]     at async /vercel/path0/node_modules/next/dist/export/worker.js:286:36
[12:10:38.592]     at async Span.traceAsyncFn (/vercel/path0/node_modules/next/dist/trace/trace.js:79:20)
[12:10:38.592] 
[12:10:38.592] > Build error occurred
[12:10:38.592] Error: Export encountered errors on following paths:
[12:10:38.592] 	/
[12:10:38.593]     at /vercel/path0/node_modules/next/dist/export/index.js:415:19
[12:10:38.593]     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[12:10:38.593]     at async Span.traceAsyncFn (/vercel/path0/node_modules/next/dist/trace/trace.js:79:20)
[12:10:38.593]     at async /vercel/path0/node_modules/next/dist/build/index.js:1400:21
[12:10:38.593]     at async Span.traceAsyncFn (/vercel/path0/node_modules/next/dist/trace/trace.js:79:20)
[12:10:38.593]     at async /vercel/path0/node_modules/next/dist/build/index.js:1259:17
[12:10:38.593]     at async Span.traceAsyncFn (/vercel/path0/node_modules/next/dist/trace/trace.js:79:20)
[12:10:38.593]     at async Object.build [as default] (/vercel/path0/node_modules/next/dist/build/index.js:66:29)
[12:10:38.639] error Command failed with exit code 1.
[12:10:38.640] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[12:10:38.662] Error: Command "yarn run build" exited with 1
[12:10:38.995] BUILD_UTILS_SPAWN_1: Command "yarn run build" exited with 1

```

### Additional information

These are the results when fetching with Axios. Instead of the data, it sends back the HTML code of the default 404 page, while on [Preview sends back a JSON](https://dev.theokondak.com/api/getSettings). I perform several other requests successfully, in the same manner as this one (the only difference is that this one has a more complex groq query) without any issues.

The Preview deployment: https://dev.theokondak.com

### Example

https://github.com/TheoKondak/portolio-minimal-reproduction

**Comment by amyegan** [comment, marked as answer]
Hi @TheoKondak! It looks like you're trying to fetch from a serverless function at build time. The 404 error is because the function doesn't exist yet. You could either change `getStaticProps` to `getServerSideProps` or call Sanity more directly to get the info you need in `getStaticProps`.

**Reply by TheoKondak** [reply]
Thanks for your reply @amyegan. I will read a bit more about `getStaticProps` and `getServerSideProps` to see how they work. Thanks for pointing me in the right direction. Still, I can't understand though, why the other requests work even though they are implemented in the same way.

Meanwhile, I was fooling around trying to understand the issue and tried fetching the data from the Preview environment, and since it was working there, the fetch request worked on my Production as well and managed to build. After that, I was able to point the fetch request back to the Production environment and I could deploy it afterward. From what you said, I guess, since it managed to build successfully once, then the function was present, and thus was able to build again to the right domain this time.
